### ISSUE TITLE
Crash when calling 'getlog io' after about 15min

### ISSUE THREAD

**Issue by SimonVrouwe** [ElementsProject/lightning#2467]
### Issue and Steps to Reproduce
I wanted to investigate another unexpected channel close (similar to  #2462) by looking at `io` log.
So I redirected `getlog io > logfile30_io` in the shell and for about 15min it was processing and then crashed (see log below).

Restarting c-lightning and calling `getlog io` crashes it again after about 15min.

### `getinfo` output
```
{
  "id": "02682020a37a3b5876b16bffbf0ac56d7f379695a5d273cf16ed0f35f18ab2dc98", 
  "alias": "LookBusy4", 
  "color": "026820", 
  "num_peers": 5, 
  "num_pending_channels": 0, 
  "num_active_channels": 4, 
  "num_inactive_channels": 1, 
  "address": [
    {
      "type": "ipv4", 
      "address": "x.x.x.x", 
      "port": 9735
    }
  ], 
  "binding": [
    {
      "type": "ipv4", 
      "address": "192.168.1.166", 
      "port": 9735
    }
  ], 
  "version": "v0.6.3rc1-276-g1dcc482", 
  "blockheight": 566847, 
  "network": "bitcoin", 
  "msatoshi_fees_collected": 0, 
  "fees_collected_msat": "0msat"
}
```

```
2019-03-13T05:59:00.701Z lightning_gossipd(22982): Received channel_update for channel 561018x1565x0/0 now ACTIVE was ACTIVE (from subdaemon)
2019-03-13T05:59:01.464Z lightningd(22970):jcon fd 26: Log pruned 121726 entries (mem 20971585 -> 3939838)
2019-03-13T05:59:02.261Z lightningd(22970):jcon fd 26: Log pruned 142045 entries (mem 20971540 -> 9680948)
... (snip)
2019-03-13T06:13:59.488Z lightningd(22970):jcon fd 26: Log pruned 148175 entries (mem 20971568 -> 10467626)
2019-03-13T06:14:00.206Z lightningd(22970):jcon fd 26: Log pruned 148176 entries (mem 20971571 -> 10469400)
2019-03-13T06:14:00.918Z lightningd(22970):jcon fd 26: Log pruned 147735 entries (mem 20971551 -> 10497907)
lightningd: Fatal signal 6 (version v0.6.3rc1-276-g1dcc482)
0x55ee3191fe17 crashdump
        common/daemon.c:40
0x7f9b98fef05f ???
        ???:0
0x7f9b98feefff ???
        ???:0
0x7f9b98ff0429 ???
        ???:0
0x55ee3196e9aa call_error
        ccan/ccan/tal/tal.c:93
0x55ee3196fd3e tal_resize_
        ccan/ccan/tal/tal.c:696
0x55ee318f7f06 membuf_tal_realloc
        lightningd/json_stream.c:46
0x55ee31965d70 membuf_prepare_space_
        ccan/ccan/membuf/membuf.c:45
0x55ee318f818e mkroom
        lightningd/json_stream.c:103
0x55ee318f834f json_stream_append_vfmt
        lightningd/json_stream.c:148
0x55ee318f8493 json_stream_append_fmt
        lightningd/json_stream.c:165
0x55ee318f8658 json_start_member
        lightningd/json_stream.c:212
0x55ee318f8985 json_add_member
        lightningd/json_stream.c:273
0x55ee318f7a7a json_add_string
        lightningd/json.c:294
0x55ee318fed70 log_to_json
        lightningd/log.c:688
0x55ee318fdd0f log_each_line_
        lightningd/log.c:345
0x55ee318fee11 json_add_log
        lightningd/log.c:706
0x55ee318ff0c7 json_getlog
        lightningd/log.c:756
0x55ee318fa04b parse_request
        lightningd/jsonrpc.c:616
0x55ee318fa474 read_json
        lightningd/jsonrpc.c:714
0x55ee31960793 next_plan
        ccan/ccan/io/io.c:59
0x55ee319612af do_plan
        ccan/ccan/io/io.c:395
0x55ee319612ed io_ready
        ccan/ccan/io/io.c:405
0x55ee31962ec8 io_loop
        ccan/ccan/io/poll.c:310
0x55ee318fcb60 main
        lightningd/lightningd.c:829
0x7f9b98fdc2e0 ???
        ???:0
0x55ee318e6199 ???
        ???:0
0xffffffffffffffff ???
        ???:0
2019-03-13T06:14:01.621Z lightningd(22970): FATAL SIGNAL 6 (version v0.6.3rc1-276-g1dcc482)
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: common/daemon.c:45 (crashdump) 0x55ee3191fe6b
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: (null):0 ((null)) 0x7f9b98fef05f
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: (null):0 ((null)) 0x7f9b98feefff
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: (null):0 ((null)) 0x7f9b98ff0429
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: ccan/ccan/tal/tal.c:93 (call_error) 0x55ee3196e9aa
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: ccan/ccan/tal/tal.c:696 (tal_resize_) 0x55ee3196fd3e
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/json_stream.c:46 (membuf_tal_realloc) 0x55ee318f7f06
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: ccan/ccan/membuf/membuf.c:45 (membuf_prepare_space_) 0x55ee31965d70
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/json_stream.c:103 (mkroom) 0x55ee318f818e
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/json_stream.c:148 (json_stream_append_vfmt) 0x55ee318f834f
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/json_stream.c:165 (json_stream_append_fmt) 0x55ee318f8493
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/json_stream.c:212 (json_start_member) 0x55ee318f8658
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/json_stream.c:273 (json_add_member) 0x55ee318f8985
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/json.c:294 (json_add_string) 0x55ee318f7a7a
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/log.c:688 (log_to_json) 0x55ee318fed70
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/log.c:345 (log_each_line_) 0x55ee318fdd0f
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/log.c:706 (json_add_log) 0x55ee318fee11
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/log.c:756 (json_getlog) 0x55ee318ff0c7
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/jsonrpc.c:616 (parse_request) 0x55ee318fa04b
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/jsonrpc.c:714 (read_json) 0x55ee318fa474
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: ccan/ccan/io/io.c:59 (next_plan) 0x55ee31960793
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: ccan/ccan/io/io.c:395 (do_plan) 0x55ee319612af
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: ccan/ccan/io/io.c:405 (io_ready) 0x55ee319612ed
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: ccan/ccan/io/poll.c:310 (io_loop) 0x55ee31962ec8
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: lightningd/lightningd.c:829 (main) 0x55ee318fcb60
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: (null):0 ((null)) 0x7f9b98fdc2e0
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: (null):0 ((null)) 0x55ee318e6199
2019-03-13T06:14:01.621Z lightningd(22970): backtrace: (null):0 ((null)) 0xffffffffffffffff
Log dumped in crash.log.20190313061401
```

**Comment by gallizoltan** [comment]
Same here.
```
lightningd: Fatal signal 6 (version v0.7.0-modded)
0x437ff7 crashdump
	common/daemon.c:40
0x7fbc07ac84af ???
	???:0
0x7fbc07ac8428 ???
	???:0
0x7fbc07aca029 ???
	???:0
0x480219 call_error
	ccan/ccan/tal/tal.c:93
0x4803d6 check_bounds
	ccan/ccan/tal/tal.c:165
0x4803fd to_tal_hdr
	ccan/ccan/tal/tal.c:173
0x480494 to_tal_hdr_or_null
	ccan/ccan/tal/tal.c:186
0x48135b tal_bytelen
	ccan/ccan/tal/tal.c:632
0x413efd json_add_hex_talarr
	lightningd/json.c:322
0x41a840 log_to_json
	lightningd/log.c:691
0x419845 log_each_line_
	lightningd/log.c:345
0x41a8a0 json_add_log
	lightningd/log.c:706
0x41ab3f json_getlog
	lightningd/log.c:756
0x415d32 parse_request
	lightningd/jsonrpc.c:616
0x41612a read_json
	lightningd/jsonrpc.c:714
0x47242d next_plan
	ccan/ccan/io/io.c:59
0x472f22 do_plan
	ccan/ccan/io/io.c:395
0x472f60 io_ready
	ccan/ccan/io/io.c:405
0x474ac0 io_loop
	ccan/ccan/io/poll.c:310
0x4186c9 main
	lightningd/lightningd.c:829
0x7fbc07ab382f ???
	???:0
0x403798 ???
	???:0
0xffffffffffffffff ???
	???:0
2019-03-13T07:37:03.878Z lightningd(17598): FATAL SIGNAL 6 (version v0.7.0-modded)
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: common/daemon.c:45 (crashdump) 0x438047
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: (null):0 ((null)) 0x7fbc07ac84af
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: (null):0 ((null)) 0x7fbc07ac8428
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: (null):0 ((null)) 0x7fbc07aca029
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: ccan/ccan/tal/tal.c:93 (call_error) 0x480219
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: ccan/ccan/tal/tal.c:165 (check_bounds) 0x4803d6
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: ccan/ccan/tal/tal.c:173 (to_tal_hdr) 0x4803fd
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: ccan/ccan/tal/tal.c:186 (to_tal_hdr_or_null) 0x480494
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: ccan/ccan/tal/tal.c:632 (tal_bytelen) 0x48135b
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: lightningd/json.c:322 (json_add_hex_talarr) 0x413efd
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: lightningd/log.c:691 (log_to_json) 0x41a840
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: lightningd/log.c:345 (log_each_line_) 0x419845
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: lightningd/log.c:706 (json_add_log) 0x41a8a0
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: lightningd/log.c:756 (json_getlog) 0x41ab3f
2019-03-13T07:37:03.878Z lightningd(17598): backtrace: lightningd/jsonrpc.c:616 (parse_request) 0x415d32
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: lightningd/jsonrpc.c:714 (read_json) 0x41612a
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: ccan/ccan/io/io.c:59 (next_plan) 0x47242d
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: ccan/ccan/io/io.c:395 (do_plan) 0x472f22
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: ccan/ccan/io/io.c:405 (io_ready) 0x472f60
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: ccan/ccan/io/poll.c:310 (io_loop) 0x474ac0
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: lightningd/lightningd.c:829 (main) 0x4186c9
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: (null):0 ((null)) 0x7fbc07ab382f
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: (null):0 ((null)) 0x403798
2019-03-13T07:37:03.879Z lightningd(17598): backtrace: (null):0 ((null)) 0xffffffffffffffff
Log dumped in crash.log.20190313073703
```
I attach my crash log.

[crash.log.20190313073703.txt](https://github.com/ElementsProject/lightning/files/2960393/crash.log.20190313073703.txt)

**Comment by cdecker** [comment]
@rustyrussell could `getlog io` be logging itself and blowing up memory as we go?

I'd propose removing the in-memory logs completely, in favor of a file-based log that is always written instead of just when we don't write to `stdout`.

**Comment by SimonVrouwe** [comment]
> instead of just when we don't write to stdout.

It also crashed when writing to `stdout`, but only with `io` level.

**Comment by trueptolemy** [comment]
@SimonVrouwe  If you system is Linux? I don't know if it is related to `overcommit_memory`.
Could you please examine the value of `vm.overcommit_memory` in `/etc/sysctl.conf`?
And you can also use `grep -i commit /proc/meminfo` and have a look at the value of `CommitLimit` and `Committed_As`.
These values help to distinguish whether it is the problem of system environment.

**Comment by gallizoltan** [comment]
I have the same crash. I'm using Linux, Ubuntu 16.04 LTS.
I've checked it, but no `vm.overcommit_memory` in my `/etc/sysctl.conf`.
I've also looked for `commit` in `meminfo` as you said:
```
$ grep -i commit /proc/meminfo
CommitLimit:     8930412 kB
Committed_AS:   17412648 kB
```
And I've checked `/proc/sys/vm/overcommit_*` values:
```
$ cat /proc/sys/vm/overcommit_kbytes
0
$ cat /proc/sys/vm/overcommit_memory
0
$ cat /proc/sys/vm/overcommit_ratio
50
```
I hope this helps.

**Comment by trueptolemy** [comment]
I guess we called [`add_entry`](https://github.com/ElementsProject/lightning/blob/3e215287cdb25d0c640b4a3f9735666f58573671/lightningd/log.c#L221) and [`prune_log`](https://github.com/ElementsProject/lightning/blob/3e215287cdb25d0c640b4a3f9735666f58573671/lightningd/log.c#L107) frequently and manufactured lots of `memory fragmentations` , which leads to `realloc()` failed.
But I'm not sure...

**Comment by trueptolemy** [comment]
Hi! @gallizoltan When your net connection lagged, did your clightning crash because of `sunc error`? 
And when you attached your log by ` getlog io` , did the crash case happen again?

**Comment by gallizoltan** [comment]
Hi! @trueptolemy I think connection lagging issue #2462  is not related here.
When my connection lagged, my c-lightning did not crash with `sync error`. Instead, I shut it down temporarily to avoid more problems. (I'm not sure shutting down made the situation better or worse.)

When `sync error` happened, it did not come to my mind to make a `getlog io`. I tried `getlog io` about one day later, and experienced the crash.

**Comment by SimonVrouwe** [comment]
Trying to find the commit that introduces the crash, my findings so far:

feb 18  c49e9ed4e5adebef163df267e9db5a3819c810d8 still ok
feb 21 f622ffb08783ebd059cb0b7b4ee1cdf3e75e2c18 NOT ok

So some in between those?

**Comment by SimonVrouwe** [comment]
Found the cullprit 53423e8a55846f405a9d36900616bc1b7137656f, which re-enables IO logging JSON output. @rustyrussell Well I guess json adds something to its log that upsets `getlog io` when called?

**Comment by SimonVrouwe** [comment]
Connecting gdb and printing `log->lr->mem_used` after node is started ends up at arround 15k, but when calling `getlog io` it blows up and keeps increasing, until it hits 20M after 15min and crashes. Perhaps we are logging the result of `getlog io` (itself) in a loop?

**Comment by trueptolemy** [comment]
@SimonVrouwe  You inspired me.
After we called `getlog io`, every time when clightning add a `log_entry` record into `json_stream`, 4 or 5 additional `log_entry`s will be generated in [`log_to_json()`](https://github.com/ElementsProject/lightning/blob/426b22fdcb360aa6ee8264991c316d858acf89f6/lightningd/log.c#L661) :

```
static void log_to_json(unsigned int skipped,
			struct timerel diff,
			enum log_level level,
			const char *prefix,
			const char *log,
			const u8 *io,
			struct log_info *info)
{
	info->num_skipped += skipped;

	if (level < info->level) {
		info->num_skipped++;
		return;
	}

	add_skipped(info);

	json_object_start(info->response, NULL);
	json_add_string(info->response, "type",
			level == LOG_BROKEN ? "BROKEN"
			: level == LOG_UNUSUAL ? "UNUSUAL"
			: level == LOG_INFORM ? "INFO"
			: level == LOG_DBG ? "DEBUG"
			: level == LOG_IO_IN ? "IO_IN"
			: level == LOG_IO_OUT ? "IO_OUT"
			: "UNKNOWN");
	json_add_time(info->response, "time", diff.ts);
	json_add_string(info->response, "source", prefix);
	json_add_string(info->response, "log", log);
	if (io)
		json_add_hex_talarr(info->response, "data", io);

	json_object_end(info->response);
}
```

Here, the function: `json_add_string()`, `json_add_time()` and `json_add_hex_talarr()` all will add new `log_entry`s.

That's why the number of `log_entry` increased so fast. These lead to that `log pruned` was called frequently.

Note we set `log->lr->max_mem` 20M. 
When `log->lr->mem_used` hit over 20M, `log pruned()` will called and about(maybe less than) 50% `log_entry`s will be deleted. So `log->lr->mem_used` will jump to ~10M.

But I think the reason of your crash is that the memory used by `json_stream` increased too big and too fast. `json_getlog()`(the function called by `getlog io`) needs add all `log_entry` with io level into `json_stream`struct by calling `log_to_json()` for every piece of `log_entry`. We use [`membuf_prepare_space_()`](https://github.com/ElementsProject/lightning/blob/426b22fdcb360aa6ee8264991c316d858acf89f6/ccan/ccan/membuf/membuf.c#L41) to expand the buffer of `json_stream`. And every time we expand buffer at least double.
In the process of outputting the io log, the number of `log_entry` grows at 4~5 times each time(maybe slower because of `log_pruned`), and the size of `json_stream`'s buffer grows at twice the rate. Finally, `tal_resize_()`(in fact, `realloc()`) failed because of too big size of `json_stream`'s buffer. :(

But on the other hand, I have no idea with @gallizoltan 's crash now. It seems different and I don't konw how to reproduct it. :(

**Comment by niftynei** [comment]
there's a fix for this in #2490, see https://github.com/ElementsProject/lightning/pull/2490/commits/ecdc12f857f99e55ea6641cc174cc7284a4d53a4

**Comment by SimonVrouwe** [comment]
> there's a fix for this in #2490, see ecdc12f

I tested that and at least it doesn't crash any more! But the result still blows up with every invocation of `getlog io`, even with a clean node that is doing nothing and not connected to anything except a regtest bitcoind.

When calling `getlog io` for the 2nd time, it includes also one big entry with (hex) data field that seems to decode to the entire result of the previous call to `getlog io`.

**Comment by cdecker** [comment]
Yeah, we might want to stop logging IO when accessing the IO log otherwise we'll see this doubling on every call.
